File: app/src/main/java/me/huidoudour/event/ui/SettingsActivity.java
package me.huidoudour.event.ui;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.card.MaterialCardView;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import me.huidoudour.event.R;
import me.huidoudour.event.data.DataImportExportHelper;
import me.huidoudour.event.data.EventRepository;
import me.huidoudour.event.utils.LocaleHelper;
import me.huidoudour.event.utils.ThemeHelper;
import me.huidoudour.event.utils.ViewModeHelper;

public class SettingsActivity extends AppCompatActivity {

    private EventViewModel viewModel;
    private EventRepository repository;
    private DataImportExportHelper dataHelper;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void attachBaseContext(Context newBase) {
        // 应用语言设置
        super.attachBaseContext(LocaleHelper.applyLanguage(newBase));
    }

    // 导出文件选择器
    private final ActivityResultLauncher<String> exportFileLauncher =
        registerForActivityResult(
            new ActivityResultContracts.CreateDocument("application/json"),
            uri -> {
                if (uri != null) {
                    executor.execute(() -> {
                        boolean success = dataHelper.exportDataToUri(repository, uri);
                        mainHandler.post(() -> {
                            if (success) {
                                Toast.makeText(this, R.string.export_success, Toast.LENGTH_SHORT).show();
                            } else {
                                Toast.makeText(this, R.string.no_data_to_export, Toast.LENGTH_SHORT).show();
                            }
                        });
                    });
                }
            }
        );

    // 导入文件选择器
    private final ActivityResultLauncher<String> importFileLauncher =
        registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    showImportConfirmDialog(uri);
                }
            }
        );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        // 在 onCreate 开始时初始化主题
        ThemeHelper.initTheme(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        viewModel = new ViewModelProvider(this, new EventViewModel.Factory(getApplication()))
                .get(EventViewModel.class);
        repository = viewModel.getRepository();
        dataHelper = new DataImportExportHelper(this);

        setupToolbar();
        setupExportData();
        setupImportData();
        setupLanguageSettings();
        setupThemeSettings();
        setupDataDisplayMode();
        setupAboutDeveloper();
    }

    private void setupToolbar() {
        com.google.android.material.appbar.MaterialToolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());
    }

    /** 导出数据 */
    private void setupExportData() {
        MaterialCardView cardExportData = findViewById(R.id.card_export_data);
        cardExportData.setOnClickListener(v ->
            exportFileLauncher.launch("events_backup_" + System.currentTimeMillis() + ".json")
        );
    }

    /** 导入数据 */
    private void setupImportData() {
        MaterialCardView cardImportData = findViewById(R.id.card_import_data);
        cardImportData.setOnClickListener(v ->
            importFileLauncher.launch("application/json")
        );
    }

    /** 显示导入确认对话框 */
    private void showImportConfirmDialog(Uri uri) {
        new MaterialAlertDialogBuilder(this)
            .setTitle(R.string.confirm_import)
            .setMessage(R.string.import_warning)
            .setPositiveButton(R.string.ok, (dialog, which) -> {
                executor.execute(() -> {
                    boolean success = dataHelper.importDataFromUri(repository, uri, true);
                    mainHandler.post(() -> {
                        if (success) {
                            Toast.makeText(this, R.string.import_success, Toast.LENGTH_SHORT).show();
                        } else {
                            Toast.makeText(this, R.string.import_failed, Toast.LENGTH_SHORT).show();
                        }
                    });
                });
            })
            .setNegativeButton(R.string.cancel, null)
            .show();
    }



    /** 语言设置 */
    private void setupLanguageSettings() {
        MaterialCardView cardLanguage = findViewById(R.id.card_language_settings);
        cardLanguage.setOnClickListener(v -> showLanguageDialog());
    }
    
    /** 显示语言选择对话框 */
    private void showLanguageDialog() {
        String[] languages = LocaleHelper.getSupportedLanguages();
        String[] languageNames = new String[languages.length];
        
        // 获取当前语言
        String currentLanguage = LocaleHelper.getLanguage(this);
        int checkedItem = 0;
        
        // 构建语言名称列表
        for (int i = 0; i < languages.length; i++) {
            languageNames[i] = LocaleHelper.getLanguageDisplayName(this, languages[i]);
            if (languages[i].equals(currentLanguage)) {
                checkedItem = i;
            }
        }
        
        new MaterialAlertDialogBuilder(this)
            .setTitle(R.string.select_language)
            .setSingleChoiceItems(languageNames, checkedItem, (dialog, which) -> {
                String selectedLanguage = languages[which];
                
                // 如果选择的语言和当前语言相同，不做任何操作
                if (selectedLanguage.equals(currentLanguage)) {
                    dialog.dismiss();
                    return;
                }
                
                // 保存并应用语言设置
                LocaleHelper.setLanguage(this, selectedLanguage);
                
                dialog.dismiss();
                
                // 显示Toast提示
                Toast.makeText(this, R.string.language_changed_restart, Toast.LENGTH_LONG).show();
                
                // 延迟重启App以应用新的语言设置
                new Handler(Looper.getMainLooper()).postDelayed(() -> {
                    restartApp();
                }, 500);
            })
            .setNegativeButton(R.string.cancel, null)
            .show();
    }
    
    /** 重启整个App */
    private void restartApp() {
        // 获取启动Intent
        Intent intent = getPackageManager().getLaunchIntentForPackage(getPackageName());
        if (intent != null) {
            // 添加标志以清除任务栈
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
            // 退出当前Activity
            finish();
            // 退出整个应用进程
            System.exit(0);
        }
    }

    /** 主题设置 */
    private void setupThemeSettings() {
        MaterialCardView cardTheme = findViewById(R.id.card_theme_settings);
        cardTheme.setOnClickListener(v -> showThemeDialog());
    }
    
    /** 显示主题选择对话框 */
    private void showThemeDialog() {
        int[] themes = ThemeHelper.getSupportedThemes();
        String[] themeNames = new String[themes.length];
        
        // 获取当前主题
        int currentTheme = ThemeHelper.getTheme(this);
        int checkedItem = 0;
        
        // 构建主题名称列表
        for (int i = 0; i < themes.length; i++) {
            themeNames[i] = ThemeHelper.getThemeDisplayName(this, themes[i]);
            if (themes[i] == currentTheme) {
                checkedItem = i;
            }
        }
        
        new MaterialAlertDialogBuilder(this)
            .setTitle(R.string.select_theme)
            .setSingleChoiceItems(themeNames, checkedItem, (dialog, which) -> {
                int selectedTheme = themes[which];
                
                // 如果选择的主题和当前主题相同，不做任何操作
                if (selectedTheme == currentTheme) {
                    dialog.dismiss();
                    return;
                }
                
                // 保存并应用主题设置
                ThemeHelper.setTheme(this, selectedTheme);
                
                dialog.dismiss();
            })
            .setNegativeButton(R.string.cancel, null)
            .show();
    }

    /** 数据展示模式设置 */
    private void setupDataDisplayMode() {
        MaterialCardView cardDataDisplayMode = findViewById(R.id.card_data_display_mode);
        
        cardDataDisplayMode.setOnClickListener(v -> showViewModeDialog());
    }
    
    /** 显示视图模式选择对话框 */
    private void showViewModeDialog() {
        int[] modes = {ViewModeHelper.VIEW_MODE_CARD, ViewModeHelper.VIEW_MODE_LIST};
        String[] modeNames = {
            getString(R.string.card_view),
            getString(R.string.list_view)
        };
        
        // 获取当前视图模式
        int currentMode = ViewModeHelper.getViewMode(this);
        int checkedItem = 0;
        
        // 找到当前选中项
        for (int i = 0; i < modes.length; i++) {
            if (modes[i] == currentMode) {
                checkedItem = i;
                break;
            }
        }
        
        new MaterialAlertDialogBuilder(this)
            .setTitle(R.string.data_display_mode)
            .setSingleChoiceItems(modeNames, checkedItem, (dialog, which) -> {
                int selectedMode = modes[which];
                
                // 如果选择的模式和当前模式相同，不做任何操作
                if (selectedMode == currentMode) {
                    dialog.dismiss();
                    return;
                }
                
                // 保存视图模式设置
                ViewModeHelper.setViewMode(this, selectedMode);
                
                dialog.dismiss();
                
                // 显示Toast提示
                android.widget.Toast.makeText(this, R.string.view_mode_changed, android.widget.Toast.LENGTH_SHORT).show();
            })
            .setNegativeButton(R.string.cancel, null)
            .show();
    }

    /** 关于开发者 */
    private void setupAboutDeveloper() {
        MaterialCardView cardAbout = findViewById(R.id.card_about_developer);
        cardAbout.setOnClickListener(v -> {
            Intent intent = new Intent(SettingsActivity.this, MeActivity.class);
            startActivity(intent);
        });
    }
}

File: app/src/main/java/me/huidoudour/event/ui/SettingsActivity.kt
package me.huidoudour.event.ui

import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.FileProvider
import androidx.core.content.edit
import androidx.lifecycle.ViewModelProvider
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.subscribeBy
import io.reactivex.rxjava3.schedulers.Schedulers
import me.huidoudour.event.MeActivity
import me.huidoudour.event.R
import me.huidoudour.event.data.DataImportExportHelper
import me.huidoudour.event.ui.theme.EventTheme
import me.huidoudour.event.util.BaseActivity
import me.huidoudour.event.util.ThemeHelper
import java.io.File

class SettingsActivity : BaseActivity() {

    private lateinit var viewModel: EventViewModel
    private lateinit var dataHelper: DataImportExportHelper
    // RxKotlin / RxJava 统一管理导入/导出异步任务，onDestroy 时自动释放
    private val disposables = CompositeDisposable()

    // 使用 mutableStateOf 驱动 Compose 重组，避免不必要的 recreate() 导致 UI 抖动
    private var isDarkTheme by mutableStateOf(false)
    private var isAscending by mutableStateOf(false)

    private val exportFileLauncher = registerForActivityResult(
        ActivityResultContracts.CreateDocument("application/json")
    ) { uri ->
        if (uri != null) {
            Single.fromCallable {
                dataHelper.exportDataToUri(viewModel.getRepository(), uri)
            }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                    onSuccess = { success ->
                        if (success) {
                            Toast.makeText(this, R.string.export_success, Toast.LENGTH_SHORT).show()
                        } else {
                            Toast.makeText(this, R.string.no_data_to_export, Toast.LENGTH_SHORT).show()
                        }
                    },
                    onError = { e -> Log.e(TAG, "exportDataToUri failed", e) }
                )
                .addTo(disposables)
        }
    }

    /** 导入：FileManager ACTION_GET_CONTENT 或系统 SAF */
    private val importFileLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            result.data?.data?.let { uri ->
                Log.d(TAG, "importFileLauncher: uri=$uri")
                showImportConfirmDialog(uri)
            }
        }
    }

    /** 导出到 FileManager：ACTION_SEND 结果处理 */
    private val exportToFileManagerLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            Toast.makeText(this, R.string.export_success, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        // enableEdgeToEdge 必须在 super.onCreate() 之前调用
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)

        viewModel = ViewModelProvider(
            this,
            EventViewModel.Factory(application)
        )[EventViewModel::class.java]
        dataHelper = DataImportExportHelper(this)

        // 初始化状态
        isDarkTheme = ThemeHelper.getTheme(this) == ThemeHelper.THEME_DARK ||
                (ThemeHelper.getTheme(this) == ThemeHelper.THEME_SYSTEM && isNightMode())
        isAscending = getSortPrefs().getBoolean("sort_ascending", false)

        applyContent()
    }

    override fun onDestroy() {
        disposables.dispose()
        super.onDestroy()
    }

    /**
     * 设置 Compose 内容。
     * 使用 key(isDarkTheme) 驱动主题切换时 Compose 树的完全重建。
     */
    private fun applyContent() {
        setContent {
            key(isDarkTheme) {
                EventTheme(darkTheme = isDarkTheme) {
                    SettingsScreenContent(
                        onBack = { finish() },
                        onAboutDeveloper = {
                            startActivity(Intent(this, MeActivity::class.java))
                        },
                        onOpenLicenses = {
                            startActivity(Intent(this, OpenSourceLicensesActivity::class.java))
                        },
                        onExport = {
                            val installed = isFileManagerInstalled()
                            Log.d(TAG, "onExport: isFileManagerInstalled=$installed")
                            if (installed) {
                                // 优先使用 FileManager 保存（Maybe：null 表示无数据可导出，走 onComplete）
                                Maybe.fromCallable {
                                    writeExportJsonToCache(viewModel.getRepository())
                                }
                                    .subscribeOn(Schedulers.io())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .subscribeBy(
                                        onSuccess = { file -> launchExportToFileManager(file) },
                                        onError = { e -> Log.e(TAG, "writeExportJsonToCache failed", e) },
                                        onComplete = {
                                            Toast.makeText(
                                                this@SettingsActivity,
                                                R.string.no_data_to_export,
                                                Toast.LENGTH_SHORT
                                            ).show()
                                        }
                                    )
                                    .addTo(disposables)
                            } else {
                                exportFileLauncher.launch("events_backup_${System.currentTimeMillis()}.json")
                            }
                        },
                        onImport = {
                            val installed = isFileManagerInstalled()
                            Log.d(TAG, "onImport: isFileManagerInstalled=$installed")
                            val intent = if (installed) {
                                Intent(Intent.ACTION_GET_CONTENT).apply {
                                    type = "application/json"
                                    setClassName(FILE_MANAGER_PACKAGE, "$FILE_MANAGER_PACKAGE.MainActivity")
                                }
                            } else {
                                Intent(Intent.ACTION_GET_CONTENT).apply {
                                    type = "application/json"
                                    addCategory(Intent.CATEGORY_OPENABLE)
                                }
                            }
                            Log.d(TAG, "onImport: launching intent=$intent")
                            try {
                                importFileLauncher.launch(intent)
                            } catch (e: Exception) {
                                Log.e(TAG, "onImport: launch failed", e)
                                Toast.makeText(this@SettingsActivity, getString(R.string.cannot_launch_file_picker, e.message), Toast.LENGTH_SHORT).show()
                            }
                        },
                        isAscending = isAscending,
                        onSortOrderChanged = { ascending ->
                            isAscending = ascending
                            getSortPrefs().edit { putBoolean("sort_ascending", ascending) }
                        },
                        onThemeChanged = { theme ->
                            ThemeHelper.saveTheme(this@SettingsActivity, theme)
                            recreate()
                        },
                        onNeedsRecreate = {
                            Toast.makeText(this, R.string.settings_applied, Toast.LENGTH_SHORT).show()
                            recreate()
                        }
                    )
                }
            }
        }
    }

    private fun isNightMode(): Boolean {
        val mode = resources.configuration.uiMode and
                android.content.res.Configuration.UI_MODE_NIGHT_MASK
        return mode == android.content.res.Configuration.UI_MODE_NIGHT_YES
    }

    private fun getSortPrefs(): SharedPreferences {
        return getSharedPreferences("sort_prefs", MODE_PRIVATE)
    }

    // =========================================================================
    //  导出：FileManager 优先
    // =========================================================================

    /** 将导出 JSON 写入 cache/shared/ 目录，返回临时文件 */
    private fun writeExportJsonToCache(repository: me.huidoudour.event.data.EventRepository): File? {
        return try {
            val events = repository.getAllEventsSync()
            if (events.isEmpty()) return null
            val json = buildExportJson(events) ?: return null
            val dir = File(cacheDir, "shared")
            if (!dir.exists()) dir.mkdirs()
            val file = File(dir, "events_backup_${System.currentTimeMillis()}.json")
            file.writeText(json, Charsets.UTF_8)
            Log.d(TAG, "writeExportJsonToCache: ${file.absolutePath} (${file.length()} bytes)")
            file
        } catch (e: Exception) {
            Log.e(TAG, "writeExportJsonToCache: failed", e)
            null
        }
    }

    /** 构建导出 JSON 字符串（与 DataImportExportHelper 格式一致） */
    private fun buildExportJson(events: List<me.huidoudour.event.data.Event>): String? {
        if (events.isEmpty()) return null
        // 每个导出任务只创建一个日期格式实例，避免在循环中反复 new
        val dateFormat = java.text.SimpleDateFormat("yyyy-MM-dd HH:mm", java.util.Locale.US)
        val sb = StringBuilder("[\n")
        events.forEachIndexed { index, event ->
            if (index > 0) sb.append(",\n")
            sb.append("  {\n")
            sb.append("    \"事件标题\": \"${event.title.replace("\\", "\\\\").replace("\"", "\\\"")}\",\n")
            sb.append("    \"事件详情\": \"${(event.description ?: "").replace("\\", "\\\\").replace("\"", "\\\"")}\",\n")
            sb.append("    \"事件时间\": \"${dateFormat.format(java.util.Date(event.eventTime))}\",\n")
            sb.append("    \"createdAt\": ${event.createdAt},\n")
            sb.append("    \"updatedAt\": ${event.updatedAt}\n")
            sb.append("  }")
        }
        sb.append("\n]")
        return sb.toString()
    }

    /** 通过 ACTION_SEND 将临时文件分享给 FileManager */
    private fun launchExportToFileManager(file: File) {
        try {
            val uri = FileProvider.getUriForFile(
                this,
                "${packageName}.fileprovider",
                file
            )
            val intent = Intent(Intent.ACTION_SEND).apply {
                type = "application/json"
                putExtra(Intent.EXTRA_STREAM, uri)
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                setClassName(FILE_MANAGER_PACKAGE, "$FILE_MANAGER_PACKAGE.MainActivity")
            }
            Log.d(TAG, "launchExportToFileManager: ACTION_SEND uri=$uri")
            exportToFileManagerLauncher.launch(intent)
        } catch (e: Exception) {
            Log.e(TAG, "launchExportToFileManager: failed", e)
            file.delete()
            Toast.makeText(this, getString(R.string.cannot_launch_file_manager, e.message), Toast.LENGTH_SHORT).show()
        }
    }

    // =========================================================================
    //  导入：FileManager 优先
    // =========================================================================

    private fun isFileManagerInstalled(): Boolean {
        return try {
            packageManager.getPackageInfo(FILE_MANAGER_PACKAGE, 0)
            true
        } catch (_: PackageManager.NameNotFoundException) {
            false
        }
    }

    private fun showImportConfirmDialog(uri: Uri) {
        // 主题实时切换后 Activity 不重建，需要包装 Context 保证对话框深浅配色正确
        androidx.appcompat.app.AlertDialog.Builder(ThemeHelper.createNightAwareContext(this))
            .setTitle(R.string.confirm_import)
            .setMessage(R.string.import_warning)
            .setPositiveButton(R.string.ok) { _, _ ->
                Single.fromCallable {
                    dataHelper.importDataFromUri(viewModel.getRepository(), uri, true)
                }
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeBy(
                        onSuccess = { success ->
                            if (success) {
                                Toast.makeText(this, R.string.import_success, Toast.LENGTH_SHORT).show()
                            } else {
                                Toast.makeText(this, R.string.import_failed, Toast.LENGTH_SHORT).show()
                            }
                        },
                        onError = { e -> Log.e(TAG, "importDataFromUri failed", e) }
                    )
                    .addTo(disposables)
            }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    companion object {
        private const val TAG = "SettingsActivity"
        /** FileManager 文件管理器包名 */
        private const val FILE_MANAGER_PACKAGE = "me.huidoudour.file.manager"
    }
}
